linkerd/protocol/thrift/src/main/scala/com/twitter/finagle/buoyant/linkerd/ThriftTraceInitializer.scala

```diff
@@ -0,0 +1,31 @@
+package com.twitter.finagle.buoyant.linkerd
+
+import com.twitter.finagle._
+import com.twitter.finagle.tracing.{Trace, TraceInitializerFilter, Tracer}
+
+object ThriftTraceInitializer {
+  val role = TraceInitializerFilter.role
+
+  val serverModule: Stackable[ServiceFactory[Array[Byte], Array[Byte]]] =
+    new Stack.Module1[param.Tracer, ServiceFactory[Array[Byte], Array[Byte]]] {
+      val role = ThriftTraceInitializer.role
+      val description = "Ensure that there is a trace id set"
+
+      def make(_tracer: param.Tracer, next: ServiceFactory[Array[Byte], Array[Byte]]) = {
+        val param.Tracer(tracer) = _tracer
+        new ServerFilter(tracer) andThen next
+      }
+    }
+
+  class ServerFilter(tracer: Tracer)
+    extends SimpleFilter[Array[Byte], Array[Byte]] {
+
+    def apply(req: Array[Byte], service: Service[Array[Byte], Array[Byte]]) = {
+      if (!Trace.hasId)
+        Trace.letTracerAndNextId(tracer) {
```

Comment:
My thinking is that if the trace context is loaded from the request, it's at that point that the span ID should be advanced.  If that has already happened by this point we don't want to advance the span ID again.  If it hasn't happened yet then the trace ID generated here will simply be overwritten by the trace loaded from the request.